Improve and extend the security checklist (#2565)
ibexa/documentation-developer: docs/infrastructure_and_maintenance/security/security_checklist.md

    Security is an ongoing process. After going live, you should pay attention to security advisories released via [your service portal](https://support.ibexa.co/), or via [Security advisories](https://developers.ibexa.co/security-advisories) if you're not a subscriber.

## [[= product_name =]]

> ### Fully-vetted admin users

Reviewer: Maybe, in the paragraph below I would remind that the user list must be kept up-to-date, anyone leaving the organization should be disable without delay, or even one changing job in the organization, should have role updated.

> ### Fully-vetted admin users

Author: "Carefully selected admin users"? Good point about the internal users.

> ### Fully-vetted admin users

Author: @adriendupuis Keeping users up to date handled in https://github.com/ibexa/documentation-developer/pull/2565/commits/377d9c8f231f0ba196ea5fc5277bb02766a37365

            - { path: ^/search, roles: ROLE_USER}
    ```

## Symfony

### `APP_SECRET`

`APP_SECRET` needs to be a strong, random, securely stored value.

- Don't use a default value like `ff6dc61a329dc96652bb092ec58981f7` or `ThisTokenIsNotSoSecretChangeIt`.

> - The secret must be secured against unwanted access. Don't commit the value to a version control system.

Reviewer: Maybe something about `.env.local` containing all the secrets, passwords, and sensible data, even `.env.local` added to `.gitignore`.

Author: Good stuff. Also, I'll need to modify the related "Secure secrets" section.

Author: ```suggestion
- The secret must be secured against unwanted access. Don't commit the value to a version control system. There are several ways of handling it, like with enviroment variables or files like `.env.local`. Files are considered more secure. If you store the secrets in files, make sure to add those files to `.gitignore` or similar, so they will never be committed to version control systems.
```

Author: Merged "Secure secrets" into this section: https://github.com/ibexa/documentation-developer/pull/2565/commits/6941f052cab1fe9d8f56c551a56a2feae940ff4d

- The secret must be long enough. 32 characters is minimum, longer is better.

!!! tip

    The following command generates a 64-character-long secure random value:

>     `php -r "print bin2hex(random_bytes(32));"`

Reviewer: To have the copy to clipboard button:

```suggestion
    ```bash
    php -r "print bin2hex(random_bytes(32));"
    ```
```

            - { path: ^/search, roles: ROLE_USER}
    ```

## Symfony

### `APP_SECRET`

`APP_SECRET` needs to be a strong, random, securely stored value.

- Don't use a default value like `ff6dc61a329dc96652bb092ec58981f7` or `ThisTokenIsNotSoSecretChangeIt`.
- The secret must be secured against unwanted access. Don't commit the value to a version control system.
- The secret must be long enough. 32 characters is minimum, longer is better.

!!! tip

    The following command generates a 64-character-long secure random value:

    `php -r "print bin2hex(random_bytes(32));"`

!!! note

    On [[= product_name_cloud =]], if `APP_SECRET` isn't set, the system sets it to [`PLATFORM_PROJECT_ENTROPY`](https://docs.platform.sh/guides/symfony/environment-variables.html#symfony-environment-variables)

### Symfony production mode

Only expose Symfony production mode openly on the internet.
Don't expose the dev mode on the internet, otherwise you may disclose things like `phpinfo` and environment variables.

> Exposing the dev mode exposes things like `phpinfo`, environment variables, and more.

Reviewer 2: Repetition of the above.

Author: Huh, that's a flaw from long ago apparently, but let's fix it. Seems we can simply delete the line.
```suggestion
```

> Exposing the dev mode exposes things like `phpinfo`, environment variables, and more.

Reviewer 3: ```suggestion
Don't expose the dev mode on the internet, otherwise you may disclose things like `phpinfo`, environment variables, and more.
```

Author: @dabrt This was resolved earlier in a different way, let me know if you still want it changed.

> For more information about securing Symfony-based systems, see [Authentication and authorisation]([[= symfony_doc =]]/security.html), [more on this subject]([[= symfony_doc =]]/security.html#learn-more), and Symfony's [secrets management system]([[= symfony_doc =]]/configuration/secrets.html).

Reviewer 3: ```suggestion
For more information about securing Symfony-based systems, see [Authentication and authorisation]([[= symfony_doc =]]/security.html), [more on this subject]([[= symfony_doc =]]/security.html#learn-more), and [secrets management system]([[= symfony_doc =]]/configuration/secrets.html), all from Symfony.
```

## PHP

### Enable `zend.exception_ignore_args` in PHP 7.4 and newer

PHP 7.4 introduced the `zend.exception_ignore_args` setting in `php.ini`.
The default value is 0 (disabled) for backwards compatibility.

> On production sites this should be set to 1 (enabled), to ensure stack traces don't include arguments passed to functions.

Reviewer 3: ```suggestion
On production sites, this should be set to 1 (enabled) to ensure that stack traces don't include arguments passed to functions.
```

Such arguments could include passwords or other sensitive information.

> You should also make sure no stack trace is ever visible to end users of production sites, though visible arguments are unsafe even if the stack traces only show up in log files.

Reviewer 3: ```suggestion
You should also make sure that no stack trace is ever visible to end users of production sites. Visible arguments are unsafe even if the stack traces only show up in log files.
```

### Disable error output from PHP

Symfony in production mode prevents exception messages from being visible to end users.
However, if Symfony fails to boot properly, such exceptions may end up being visible, including stack traces.
This can be prevented by [disabling error message output in PHP](https://www.php.net/manual/en/language.errors.basics.php).

> These `php.ini` configuration values should be used on production sites.

Reviewer 3: These? Do you mean "the following"?

Author: ```suggestion
The following `php.ini` configuration values should be used on production sites.
```

When using [[= product_name_cloud =]], the same settings can be configured in [[= product_name =]]'s `.platform.app.yaml` file.

```ini
display_errors          = Off
display_startup_errors  = Off
```

### Other PHP settings

Consider what other security related settings are relevant for your needs.
The [OWASP PHP Configuration Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/PHP_Configuration_Cheat_Sheet.html) contains several recommendations, but be aware that they may be out of date as they don't mention PHP 8.

For more information, see [PHP's own security manual](https://www.php.net/manual/en/security.php).

## Web server

### Block execution of scripts in `var` directory

> Make sure the web server blocks the execution of PHP files and other scripts in the `var` directory.

Reviewer 3: ```suggestion
Make sure that the web server blocks the execution of PHP files and other scripts in the `var` directory.
```

TLS 1.3 is as of December 2024 [supported by ca. 97% of global internet users](https://caniuse.com/tls1-3).
If you need to support Internet Explorer or old versions of other browsers, you can disable TLS 1.1 and older, leaving 1.2 and 1.3 enabled.

When using [[= product_name_cloud =]], you can [set the minimum TLS version in .platform/routes.yaml](https://docs.platform.sh/define-routes/https.html#enforce-tls-13).

### Enable HTTP Strict Transport Security (HSTS)

HSTS forces clients to always communicate with your site over HTTPS.
[Most browsers support this](https://caniuse.com/stricttransportsecurity), and there is no downside for browsers that don't.
Read the requirements and instructions at [hstspreload.org](https://hstspreload.org/) before you enable HSTS.
Make sure to also include subdomains using the `includeSubDomains` setting.

When using [[= product_name_cloud =]], you can [configure HSTS in .platform/routes.yaml](https://docs.platform.sh/define-routes/https.html#enable-http-strict-transport-security-hsts).

## Domain

### Enable Domain Name System Security Extensions (DNSSEC)

DNSSEC is a DNS feature that authenticates responses to DNS requests.

> It protects against DNS poisoning attacks.
> This is when an attacker manipulates the reponses to DNS requests with the goal of directing users to an IP address the attacker controls.

Author: Vale suggestion:
```suggestion
It protects against DNS poisoning attacks, which is when an attacker manipulates the reponses to DNS requests with the goal of directing users to an IP address the attacker controls.
```

Author: Pending reviewers approval.

Enabling DNSSEC involves creating the DNSSEC records in your domain, activating DNSSEC with your domain registrar, and enabling DNSSEC signature validation on all DNS servers.
[Read more on DNSSEC at ICANN](https://www.icann.org/resources/pages/dnssec-what-is-it-why-important-2019-03-05-en).

### Enable domain update/delete protection

Domain update/delete protection is a DNS setting that makes it harder for an attacker to take over a domain from the real owner, or hinder availability for users.

> This is configured with your domain registrar.

Author: Vale suggestion:
```suggestion
You can enable this protection at your domain registrar's site.
```

If you're using MySQL/MariaDB, use the UTF8MB4 database character set and related collation.
The older UTF8 can lead to truncation with 4-byte characters, like some emoji, which may have unpredictable side effects.

See [Change from UTF8 to UTF8MB4](update_db_to_2.5.md#change-from-utf8-to-utf8mb4).

### Secure access

Secure the database access with strong passwords, keys, firewall, encryption in transit, encryption at rest etc. as needed.
When using [[= product_name_cloud =]], the provider handles this.

### Limit database rights

Optionally, ensure that the database user used by the web app only has permissions to do the operations needed by [[= product_name =]].
The Data Definition Language (DDL) commands (create, alter, drop, truncate, comment) are only needed for installing and upgrading [[= product_name =]], and not for running it.
Not granting these rights to web app users reduces the damage that can result from a security breach.

## Underlying stack

To avoid exposing your application to any DDOS vulnerabilities or other yet unknown security threats, make sure that you do the following:

- Avoid exposing servers on the open internet when not strictly required.

> - Ensure any servers, services, ports and virtual hosts that were opened for testing purposes are shut down before going live.

Author: Oxford comma 🧐 
```suggestion
- Ensure any servers, services, ports, and virtual hosts that were opened for testing purposes are shut down before going live.
```

CAA allows domain owners to specify which Certificate Authorities (CAs) are permitted to issue SSL/TLS certificates for their domain.
This prevents attackers from having certificates issued for domains they don't own, hindering some types of attack.
CAA is configured in your DNS zone file.

## Database

### Use UTF8MB4 with MySQL/MariaDB

If you're using MySQL/MariaDB, use the UTF8MB4 database character set and related collation.
The older UTF8 can lead to truncation with 4-byte characters, like some emoji, which may have unpredictable side effects.

See [Change from UTF8 to UTF8MB4](update_db_to_2.5.md#change-from-utf8-to-utf8mb4).

### Secure access

Secure the database access with strong passwords, keys, firewall, encryption in transit, encryption at rest etc. as needed.
When using [[= product_name_cloud =]], the provider handles this.

### Limit database rights

Optionally, ensure that the database user used by the web app only has permissions to do the operations needed by [[= product_name =]].
The Data Definition Language (DDL) commands (create, alter, drop, truncate, comment) are only needed for installing and upgrading [[= product_name =]], and not for running it.

> Not granting these rights to web app users reduces the damage that can result from a security breach.

Reviewer: So, what at the rights to give to web app users to have the needed ones and only the needed ones? (see #2356)

Author: @adriendupuis This isn't documented AFAIK. Ref. the doc that only says "Before executing the command make sure that the database user has sufficient permissions."
https://doc.ibexa.co/en/latest/getting_started/install_ibexa_dxp/#create-a-database

Incidentally, I found that the next paragraph doesn't belong there. PR to move it: https://github.com/ibexa/documentation-developer/pull/2569

Anyway, I removed the section, and instead added a new Monitor section at the end, about monitoring the various part of the system. See https://github.com/ibexa/documentation-developer/pull/2565/commits/26d5a78bbefe56e76027093f57f525353420657f

- `Permissions-Policy` - limits what features the browser can use, such as fullscreen, notifications, location, camera, or microphone.
For example, if someone succeeds in injecting their JavaScript into your site, this header prevents them from using those features to attack your users.

### Disable weak cipher suites in TLS

Consider blocking the use of TLS 1.2 and older versions.
The newer TLS 1.3 does not include the weaker cipher suites that are included in 1.2 and older.

> Removing them means attackers can't attempt to force other users to use weak ciphers and eavesdrop on their communications.

Reviewer 3: ```suggestion
Removing them means that attackers can't attempt to force other users to use weak ciphers and eavesdrop on their communications.
```

> TLS 1.3 is as of December 2024 [supported by ca. 97% of global internet users](https://caniuse.com/tls1-3).

Reviewer 3: ```suggestion
TLS 1.3 is as of December 2024 [supported by ca. 97% of global internet users](https://caniuse.com/tls1-3).
```
I'd put "As of December 2024," in the first place in the sentence.

If you need to support Internet Explorer or old versions of other browsers, you can disable TLS 1.1 and older, leaving 1.2 and 1.3 enabled.

> When using [[= product_name_cloud =]], you can [set the minimum TLS version in .platform/routes.yaml](https://docs.platform.sh/define-routes/https.html#enforce-tls-13).

Reviewer 3: ```suggestion
When using [[= product_name_cloud =]], you can [set the minimum TLS version in `.platform/routes.yaml`](https://docs.platform.sh/define-routes/https.html#enforce-tls-13).
```

> ### Enable HTTP Strict Transport Security (HSTS)

Reviewer 4: Or actually maybe I misunderstood it and it's ok to enforce HSTS, because the purge request is sent from the back office to Varnish, not the other way🤔 I'm not sure right now

> ### Enable HTTP Strict Transport Security (HSTS)

Author: Nice catch. We can say that this should apply between the client and the public-facing proxy, if any, otherwise the webserver. I'll come up with a sentence or warning box about it.

> ### Enable HTTP Strict Transport Security (HSTS)

Reviewer 4: It's safer for me to ping @vidarl here, he knows better what's the recommended approach 😄

> ### Enable HTTP Strict Transport Security (HSTS)

Reviewer 5: These settings are related to what happends between client and public-facing proxy, yes.

Even with both https and HSTS enabled, varnish and app container will still be reachable to each other over plain http

> ### Enable HTTP Strict Transport Security (HSTS)

Author: Thanks, handled in https://github.com/ibexa/documentation-developer/pull/2565/commits/7eb8d61d5482fdcf3f1dafc3c0534b4a847bc230

HSTS forces clients to always communicate with your site over HTTPS.
[Most browsers support this](https://caniuse.com/stricttransportsecurity), and there is no downside for browsers that don't.
Read the requirements and instructions at [hstspreload.org](https://hstspreload.org/) before you enable HSTS.

> Make sure to also include subdomains using the `includeSubDomains` setting.

Reviewer 3: ```suggestion
Make sure to also include subdomains by means of the `includeSubDomains` setting.
```

> When using [[= product_name_cloud =]], you can [configure HSTS in .platform/routes.yaml](https://docs.platform.sh/define-routes/https.html#enable-http-strict-transport-security-hsts).

Reviewer 3: ```suggestion
When using [[= product_name_cloud =]], you can [configure HSTS in `.platform/routes.yaml`](https://docs.platform.sh/define-routes/https.html#enable-http-strict-transport-security-hsts).
```

## Domain

### Enable Domain Name System Security Extensions (DNSSEC)

DNSSEC is a DNS feature that authenticates responses to DNS requests.
It protects against DNS poisoning attacks, which is when an attacker manipulates the reponses to DNS requests with the goal of directing users to an IP address the attacker controls.
Enabling DNSSEC involves creating the DNSSEC records in your domain, activating DNSSEC with your domain registrar, and enabling DNSSEC signature validation on all DNS servers.

> [Read more on DNSSEC at ICANN](https://www.icann.org/resources/pages/dnssec-what-is-it-why-important-2019-03-05-en).

Reviewer 3: ```suggestion
[Read more on DNSSEC on ICANN's website](https://www.icann.org/resources/pages/dnssec-what-is-it-why-important-2019-03-05-en).
```

### Enable domain update/delete protection

Domain update/delete protection is a DNS setting that makes it harder for an attacker to take over a domain from the real owner, or hinder availability for users.
You can enable this protection at your domain registrar's site.
Log in to their site to enable these protection settings and save the new configuration.

### Enable Certificate Authority Authorization (CAA)

CAA allows domain owners to specify which Certificate Authorities (CAs) are permitted to issue SSL/TLS certificates for their domain.
This prevents attackers from having certificates issued for domains they don't own, hindering some types of attack.
CAA is configured in your DNS zone file.

## Database

### Use UTF8MB4 with MySQL/MariaDB

If you're using MySQL/MariaDB, use the UTF8MB4 database character set and related collation.
The older UTF8 can lead to truncation with 4-byte characters, like some emoji, which may have unpredictable side effects.

See [Change from UTF8 to UTF8MB4](update_db_to_2.5.md#change-from-utf8-to-utf8mb4).

### Secure access

> Secure the database access with strong passwords, keys, firewall, encryption in transit, encryption at rest etc. as needed.

Reviewer 3: ```suggestion
Secure the database access with strong passwords, keys, firewall, encryption in transit, encryption at rest, and so on, as needed.
```
